### ISSUE TITLE
`vim.highlight.range` doesn't result into latest added highlight

### ISSUE THREAD

**Issue by echasnovski** [neovim/neovim#17358]
### Neovim version (nvim -v)

NVIM v0.7.0-dev+1048-gdba1df635

### Vim (not Nvim) behaves the same?

No, Vim doesn't have Lua functions

### Operating system/version

Ubuntu 20.04

### Terminal name/version

kitty 0.23.1

### $TERM environment variable

xterm-kitty

### Installation

nvim.appimage

### How to reproduce the issue

1. `nvim --clean -u NONE`.
2. Type `abc`.
3. `:lua vim.highlight.range(0, 1, 'Comment', {0, 0}, {0, -1})` (whole line should be highlighted as comment)
4. `:lua vim.highlight.range(0, 1, 'Delimiter', {0, 0}, {0, -1})` (whole line should be highlighted as delimiter)
5. `:lua vim.highlight.range(0, 1, 'Comment', {0, 0}, {0, -1})`

### Expected behavior

Whole line again should be highlighted as comment.

### Actual behavior

Highlighting of whole doesn't change.

Notes:
- If last highlighting is done with bigger priority ( `:lua vim.highlight.range(0, 1, 'Comment', {0, 0}, {0, -1}, 'v', false, 51)`), everything is as expected.
- Everything works as expected in 0.6.0.

**Comment by zeertzjq** [comment]
Bisected to <https://github.com/neovim/neovim/commit/95ab979fde66d8f9f97fceb943bfe9422739a0f8>.

**Comment by shadmansaleh** [comment]
This is actually quite funny . The default priority is too low [here](https://github.com/neovim/neovim/blob/046950f6309070d98ab5a3deeff9d00e079ccdd4/runtime/lua/vim/highlight.lua#L32) . Treesitter uses priority 100 . The default one to 200 or something 🤣

**Comment by clason** [comment]
No, I don't think that's it; I can reproduce this without any tree-sitter highlighting. The problem is that a second _identical_ highlight is not applied again (and hence doesn't overwrite a _different_ highlight applied in the mean time).

**Comment by shadmansaleh** [comment]
> The problem is that a second identical highlight is not applied again (and hence doesn't overwrite a different highlight applied in the mean time)

You're right . It should overright older marks with same priority . But the priority level should be increased too  otherwise this is not expected to work at all right as ts has higher priority .

**Comment by clason** [comment]
That is very much by design. Setting the right priority is the responsibility of whoever calls this function, depending on the exact use case.

**Comment by shadmansaleh** [comment]
Reverting change to this [line](https://github.com/neovim/neovim/blob/046950f6309070d98ab5a3deeff9d00e079ccdd4/src/nvim/marktree.c#L116) seems to fix it . Still trying to figure out why .
```diff
 diff --git a/src/nvim/marktree.c b/src/nvim/marktree.c
index 918db8b76..f8365f7c1 100644
--- a/src/nvim/marktree.c
+++ b/src/nvim/marktree.c
@@ -113,7 +113,7 @@ static int key_cmp(mtkey_t a, mtkey_t b)
   }
   // NB: keeping the events at the same pos sorted by id is actually not
   // necessary only make sure that START is before END etc.
-  return mt_generic_cmp(a.flags, b.flags);
+  return mt_generic_cmp(a.id, b.id);
 }
 
 static inline int marktree_getp_aux(const mtnode_t *x, mtkey_t k, int *r)
```

**Comment by clason** [comment]
The correct fix for this is to add more book-keeping to `vim.highlight.range` to clean up outdated extmarks before applying new ones. There is no guaranteed ordering of extmarks by order of application, and relying on that assumption is wrong.

**Comment by lsvmello** [comment]
> The correct fix for this is to add more book-keeping to `vim.highlight.range` to clean up outdated extmarks before applying new ones. There is no guaranteed ordering of extmarks by order of application, and relying on that assumption is wrong.

Is this the way to go here?

I could try to implement it but, since there is `opts.priority`, should be user responsibility to correctly place the highlight order.
